kafka-source/src/test/java/org/springframework/cloud/stream/module/kafka/source/KafkaSourceIntegrationTests.java

```diff
@@ -0,0 +1,306 @@
+			Integer[] in = new Integer[]{0, 1, 2, 4, 5, 6, 8, 9, 10};
+			for (int i : in) {
+				assertTrue(strings.contains(Integer.toString(i)));
+			}
+			Integer[] excluded = new Integer[]{3, 7, 11};
+			for (int i : excluded) {
+				assertFalse(strings.contains(Integer.toString(i)));
+			}
+		}
+	}
+
+	@IntegrationTest({"partitions.bar=0,1,2,3", "initialOffsets.bar.0=5"})
+	public static class InitialOffsetForOnePartitionStartsOutsideRangeTest extends KafkaSourceIntegrationTests {
+
+		@Test
+		public void testTopicWithOnlyConfiguredPartitions() throws Exception {
+
+			//12 messages produced, only 9 are consumed as partition 0's initial offsets are outside
+			//range and not committed to zookeeper.
+			for (int i = 0; i < 12; i++) {
+				producer.send(new ProducerRecord<>(TEST_TOPIC2, i % 4,
+						"test-topic" + i,
+						Integer.toString(i)));
+			}
+
+			List<String> strings = new ArrayList<>();
+			for (int i = 0; i < 9; i++) {
+				Message<?> out = this.messageCollector.forChannel(this.source.output()).poll(10, TimeUnit.SECONDS);
+				assertNotNull(out);
+				String payload = new String(((byte[]) out.getPayload()), "UTF-8");
+				strings.add(payload);
+				assertEquals(TEST_TOPIC2, out.getHeaders().get(KafkaHeaders.TOPIC));
+			}
+
+			Message<?> out = this.messageCollector.forChannel(this.source.output()).poll(4, TimeUnit.SECONDS);
+			assertNull(out);
+
+			//Verify that we read from all the partitions and all messages are accounted for
+			Integer[] in = new Integer[]{1, 2, 3, 5, 6, 7, 9, 10, 11};
+			for (int i : in) {
+				assertTrue(strings.contains(Integer.toString(i)));
+			}
+			Integer[] excluded = new Integer[]{0, 4, 8};
+			for (int i : excluded) {
+				assertFalse(strings.contains(Integer.toString(i)));
+			}
+		}
+	}
+
+	@IntegrationTest({"partitions.bar=0,1,2,3", "initialOffsets.bar.0=0"})
+	public static class InitialOffsetForOnePartitionSetExplicitlyToBeginningTest extends KafkaSourceIntegrationTests {
+
+		@Test
+		public void testTopicWithOnlyConfiguredPartitions() throws Exception {
+
+			//12 messages are going in, all 12 should be consumed from 4 partitions.
+			for (int i = 0; i < 12; i++) {
+				producer.send(new ProducerRecord<>(TEST_TOPIC2, i % 4,
+						"test-topic" + i,
+						Integer.toString(i)));
+			}
+
+			List<String> strings = new ArrayList<>();
+			for (int i = 0; i < 12; i++) {
+				Message<?> out = this.messageCollector.forChannel(this.source.output()).poll(10, TimeUnit.SECONDS);
+				assertNotNull(out);
+				String payload = new String(((byte[]) out.getPayload()), "UTF-8");
+				strings.add(payload);
+				assertEquals(TEST_TOPIC2, out.getHeaders().get(KafkaHeaders.TOPIC));
+			}
+
+			//Verify that we read from all the partitions and all messages are accounted for
+			Integer[] in = new Integer[]{0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11};
+			for (int i : in) {
+				assertTrue(strings.contains(Integer.toString(i)));
+			}
+		}
+	}
+
+	@Configuration
+	@EnableIntegration
```

Comment:
This stuff is redundant for the configuration below.
Just chasing for some test performance. :smile: